### PR TITLE
fix(bunx): can't resolve executable for scoped packages

### LINKED ISSUE #2292: `bunx` fails to find the executable for scoped package
### What version of Bun is running?

0.5.7

### What platform is your computer?

Darwin 22.4.0 arm64 arm

### What steps can reproduce the bug?

When using `bunx` with scoped packages i.e. `@withfig/autocomplete-tools` it fails to find the binary to run

1. `bunx @withfig/autocomplete-tools --version`

### What is the expected behavior?

Example with npx
```
franzekan@Frans-M1-Macbook-Pro: /Users/franzekan
➜   npx @withfig/autocomplete-tools --version
Need to install the following packages:
  @withfig/autocomplete-tools@2.7.9
Ok to proceed? (y)
1.0.0
```

Fixed bun (still a WIP locally on my machine)
```
⬢ [Docker] ❯ bun-debug x @withfig/autocomplete-tools --version
initial_bin_name: autocomplete-tools
1.0.0
```

### What do you see instead?

```
➜   bunx @withfig/autocomplete-tools --version
error: could not determine executable to run for package @withfig/autocomplete-tools
```

### Additional information

I've fixed this so it works if the package is already installed (bun i package; bunx package), so it still breaks if you delete the cache on first run i.e.
```
bunx @foo/package -> fails
bunx @foo/package -> works (and every subsequent)
```

### PR DESCRIPTION
**Pull request by Zeko369**
Fixes #2292

- [x] Works if package is installed
```bash
bun-debug x @antfu/ni --version # ignore that this one failed as it will install it
bun-debug x @antfu/ni --version
```

- [x] Works if package needs to be installed (clear cache)
```bash
rm -rf /tmp/@antfu
bun-debug x @antfu/ni --version
```

### CODE REVIEW

**Comment by Jarred-Sumner** [comment]
Thank you for this.

Would love a test to go along with it. You can use `Bun.spawnSync` for that

Here's an example test to cargo cult from (note the usage of `bunEnv` and `bunExe`, these exist to make it use the debug or release build of Bun and set environment variables that will make it behave as expected)

https://github.com/oven-sh/bun/blob/c892ce93a50704b8302caa9c9d5de42e2cfba685/test/bun.js/install/bad-workspace.test.ts#L7-L36

File: src/cli/bunx_command.zig [hunk 1]
```diff
@@ -209,11 +209,20 @@ pub const BunxCommand = struct {
...
 
+        if (strings.startsWithChar(update_request.name, '@')) {
+            // Should be unreachable
+            if (!strings.containsChar(update_request.name, '/')) {
+                Output.prettyErrorln("<r><red>error<r><d>:<r> Invalid package name: {s}", .{update_request.name});
+                Global.exit(1);
+            }
+
+            initial_bin_name = update_request.name[strings.indexOfChar(update_request.name, '/').? + 1 ..];
+        }
```
Comment (alexlamsl) [inline src/cli/bunx_command.zig]: The following patch (on the original `bunx_command.zig`) should take care of both issues you have identified:
```diff
--- a/src/cli/bunx_command.zig
+++ b/src/cli/bunx_command.zig
@@ -204,6 +204,8 @@ pub const BunxCommand = struct {
 
         const initial_bin_name = if (strings.eqlComptime(update_request.name, "typescript"))
             "tsc"
+        else if (strings.lastIndexOfChar(update_request.name, '/')) |index|
+            update_request.name[index + 1 ..]
         else
             update_request.name;
 
@@ -365,7 +367,7 @@ pub const BunxCommand = struct {
             },
         }
 
-        absolute_in_cache_dir = std.fmt.bufPrint(&absolute_in_cache_dir_buf, "{s}/node_modules/.bin/{s}", .{ bunx_cache_dir, update_request.name }) catch unreachable;
+        absolute_in_cache_dir = std.fmt.bufPrint(&absolute_in_cache_dir_buf, "{s}/node_modules/.bin/{s}", .{ bunx_cache_dir, initial_bin_name }) catch unreachable;
 
         // Similar to "npx":
         //
```

File: test/bun.js/install/bunx.test.ts [hunk 2]
```diff
@@ -72,6 +72,25 @@ it("should output usage if no arguments are passed", async () => {
   expect(await exited).toBe(1);
 });
 
+it("should work for @scoped packages", async () => {
```
Comment (alexlamsl) [inline test/bun.js/install/bunx.test.ts:75]: consider adding:
```js
await rm(join(await realpath(tmpdir()), "@antfu"), { force: true, recursive: true });
```
here to ensure this test always run uncached.

File: test/bun.js/install/bunx.test.ts [hunk 3]
```diff
@@ -72,6 +72,25 @@ it("should output usage if no arguments are passed", async () => {
...
+    env,
+  });
+
+  expect(stderr).toBeDefined();
+  const err = await new Response(stderr).text();
+  expect(err).toBe("");
+  expect(stdout).toBeDefined();
+  const out = await new Response(stdout).text();
+  expect(out.trim().split(/\r?\n/)).toEqual(["@antfu/ni v0.20.0"]);
+  expect(await exited).toBe(0);
```
Comment (alexlamsl) [inline test/bun.js/install/bunx.test.ts]: ... you can then repeat the `spawn()` and `expect()`s here to test the cached case.

File: src/cli/bunx_command.zig (hunk 1 above)
Comment (Zeko369) [reply src/cli/bunx_command.zig]: Actually, in the meantime I got it working by doing the split logic in `getBinNameFromSubpath`, but your code is shorter/easier to understand, so I'll use that, tnx a lot 🤗

**Review by alexlamsl — APPROVED** [review]
LGTM

**Pull request merged**